src/main/java/ir/sahab/kafkaconsumer/OffsetTracker.java

```diff
@@ -119,51 +120,70 @@ public OptionalLong ack(int partition, long offset) {
     private class PartitionTracker {
         private final Map<Long /*page index*/, PageTracker> pageTrackers = new HashMap<>();
         private final SortedSet<Long /*page index*/> completedPages = new TreeSet<>();
-        private long lastConsecutivePageIndex;
+        private long lastCompletedPageIndex;
+        private long lastOpenedPageIndex;
+        private long lastTrackedOffset;
         private volatile int openPagesSize = 0;
         private volatile int completedPagesSize = 0;
 
         public PartitionTracker(int partition, long initialOffset) {
-            lastConsecutivePageIndex = offsetToPage(initialOffset) - 1;
             String partitionLabel = String.valueOf(partition);
             metricRegistry.register(LabeledMetric.name(OPEN_PAGES).label(PARTITION, partitionLabel).toString(),
                     (Gauge<Integer>) () -> openPagesSize);
             metricRegistry.register(LabeledMetric.name(COMPLETED_PAGES).label(PARTITION, partitionLabel).toString(),
                     (Gauge<Integer>) () -> completedPagesSize);
+
+            openNewPageForOffset(initialOffset);
+            lastCompletedPageIndex = offsetToPage(initialOffset) - 1;
+            lastTrackedOffset = initialOffset - 1;
         }
 
         boolean track(long offset) {
             long pageIndex = offsetToPage(offset);
-            int pageOffset = (int) (offset % pageSize);
-            PageTracker pageTracker = pageTrackers.get(pageIndex);
-            if (pageTracker == null) {
-                if (pageTrackers.size() >= maxOpenPagesPerPartition) {
+
+            // Normal case: offsets coming in order.
```

Comment:
```suggestion
            // Normal case: offsets coming in order.
```

It is not yet a normal case. When you know the page is not changed, you can tell that it is a normal case.